### PR TITLE
Implement endpoint routing

### PR DESCRIPTION
**Pull request by containscafeine**
- Implement endpoint routing
- Add docs
- Add tests

### CODE REVIEW

**Comment by containscafeine** [comment]
TODO:
- [ ] infer port values from service names
- [ ] see if we can get some envoy load balancers to work with this

File: docs/reference/load-balancer.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+The `load_balancer` attribute configures the load balancing. The following fields are supported:
+
+- `type`: Specifies the type of load balancer to use.
+- `policy`: Specifies the load balancing policy to use.
+
+### `type: kubernetes`
+The `kubernetes` type delegates load balancing to Kubernetes. Supported policies:
+- `round_robin` (default)
+
```
Comment (richarddli) [inline docs/reference/load-balancer.md]: Add: In this mode, Ambassador will route traffic directly to a Kubernetes service. Kubernetes service networking(https://kubernetes.io/docs/concepts/services-networking/) is then responsible for load balancing traffic between each of the pods.

File: docs/reference/load-balancer.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+### `type: envoy`
+The `envoy` type delegates load balancing to Envoy. Supported policies:
+- `round_robin`
+
```
Comment (richarddli) [inline docs/reference/load-balancer.md]: How do health checks work in this situation?

Comment (containscafeine) [reply docs/reference/load-balancer.md]: Kubernetes reports only healthy endpoints to Ambassador, so envoy maintains a healthy list at all time.

File: ambassador/ambassador/config/resourcefetcher.py [hunk 3]
```diff
@@ -228,3 +189,149 @@ def process_object(self, obj: dict, rkey: Optional[str]=None) -> None:
...
+            return None
+
+    def parse(self):
+        parsed = self.get_object_kind()
+        if parsed is None:
+            return None, ""
+        return parsed.parse()
+
+
+class EndpointsKind(ObjectKind):
```
Comment (kflynn) [inline ambassador/ambassador/config/resourcefetcher.py]: Should `EndpointsKind` and `ServiceKind` really just be new `ACResource` types?

File: ambassador/ambassador/utils.py [hunk 4]
```diff
@@ -100,6 +100,8 @@ def load_url_contents(logger: logging.Logger, url: str, stream2: Optional[TextIO
     stream = io.StringIO()
 
     saved = _load_url_contents(logger, url, stream, stream2=stream2)
+    if stream2:
+        stream2.write("---\n")
```
Comment (kflynn) [inline ambassador/ambassador/utils.py]: Hmmm. What's this for?

File: ambassador/ambassador/config/resourcefetcher.py [hunk 5]
```diff
@@ -105,73 +105,34 @@ def parse_yaml(self, serialization: str, k8s=False, rkey: Optional[str]=None,
...
-                    self.extract_k8s(obj)
-                else:
-                    # if not obj:
-                    #     self.logger.debug("%s: empty object from %s" % (self.location, serialization))
-
-                    self.process_object(obj, rkey=rkey)
-                    self.ocount += 1
-
-            self.pop_location()
+            self.parse_object(objects=objects, k8s=k8s, rkey=rkey, filename=filename)
```
[the remark this answers is not in the document]
Comment (kflynn) [reply ambassador/ambassador/config/resourcefetcher.py]: Y'know what? I take this back, after studying the two files side by side.

**Comment by kflynn** [comment]
We'll continue improving this, but let's get it landed for smaller PRs.

**Pull request merged**